src/server/chat.js

```diff
@@ -0,0 +1,292 @@
+const { mojangPublicKeyPem } = require('./constants')
+const crypto = require('crypto')
+const { concat } = require('../transforms/binaryStream')
+
+const lastSeenStates = {
+  REMOVED_MESSAGES: 0,
+  OUT_OF_ORDER: 1,
+  UNKNOWN_MESSAGES: 2,
+  DUPLICATED_PROFILES: 3
+}
+
+const chatVisibility = {
+  FULL: 0,
+  SYSTEM: 1,
+  HIDDEN: 2
+}
+
+module.exports = function (client, server, options) {
+  const mojangPubKey = crypto.createPublicKey(mojangPublicKeyPem)
+  const mcData = require('minecraft-data')(client.version)
+  const raise = (translatableError) => client.end(translatableError, JSON.stringify({ translate: translatableError }))
+  const throwGeneric = (message) => client.end(message, JSON.stringify({ translate: 'disconnect.genericReason', with: [{ text: message }] }))
+
+  const {
+    'enable-chat-signing': signedChat
+  } = options
+
+  client.on('chat_message', handleChatMessage)
+
+  client.on('settings', handleClientSettings)
+
+  client.on('player_chat', handleIncomingChat) // Do server-side clients emit this event?
+
+  client.on('message_acknowledgement', handleAcknowledgements)
+
+  client.on('chat_session', handleChatSession)
+
+  client.lastMessage = {
+    timestamp: 0
+  }
+
+  client.packAcknowledgements = (acknowledgements) => {
+    if (!client.signature_cache) client.signature_cache = []
+
+    return acknowledgements.map(ack => {
+      const idx = client.signature_cache.findIndex(el => el.compare(ack) === 0)
+      if (idx === -1) return { id: 0, signature: ack }
+      return { id: idx }
+    })
+  }
+
+  function handleChatSession (packet) {
+    client.session = {
+      uuid: packet.sessionUUID,
+      index: 0
+    }
+
+    const publicKey = crypto.createPublicKey({ key: packet.publicKey, format: 'der', type: 'spki' })
```

Comment:
Pretty sure the public key should already be in DER format, 3 lines here should not be needed

Reply:
actually, nevermind -- nodejs assumes PEM by default in verify https://github.com/nodejs/node/blob/4830a6cf2aa14144a1c36ae92dac7eed66fb8c13/lib/internal/crypto/keys.js#L520

Reply:
> Pretty sure the public key should already be in DER format, 3 lines here should not be needed

Public key provided by client is padded. Signature is created with an unpadded key

Reply:
Yeah, probably a good idea to do this in the client also.